### PR TITLE
Add Continuous integration to automatically generate PDF and diff-pdfs using circleCI

### PR DESCRIPTION
**Pull request by yfarjoun**
This PR adds a continuous integration option that creates the PDFs and serves them up for perusal. 

This informs circle-ci to make all the pdfs and publish them to an accessible location.
See https://circleci.com/gh/samtools/hts-specs for build information.

### CODE REVIEW

**Comment by jmarshall** [comment]
I was anticipating that you would bring this up in the meeting just now :smile:

I have some comments. But if you're presenting this as a PR potentially to be merged, before looking at it I would like to see it presented as slightly more polished single commit with a descriptive commit message, rather than including all the false starts and CI iterations.

**Comment by yfarjoun** [comment]
Yeah...I was late to the meeting and then there wasn't a good time to interject. I'll clean this up.

**Comment by yfarjoun** [comment]
@jmarshall here you go. By the way, was everyone getting build notifications (from circle-ci) over the past week, or just me?

**Comment by jkbonfield** [comment]
I haven't had emails, but I have seen lots of "x"s next to hts-specs issues where circle-ci had failed.

I like the concept of this as keeping them up to date whenever anything is merged to master seems logical.  Does this also automatically update pdfs for PRs so we can see the effect right away?

**Comment by jmarshall** [comment]
No emails, so presumably they are only sent to the person who instigated the build. Also either it's remarkably non-obvious where the uploaded artifacts are, or they're only available to you or to people who have logged in to circle-ci. The latter would make this less interesting.

Perhaps you could explain what you find awkward about the existing process for making PDFs. (BTW I've committed a note about the tricks I use for facilitating working on these PDFs — see d434cc51615705a5de1599f6d26910cdd74c0f2b and _MAINTAINERS.md_.) Is it just that you have to make PDFs at all?

What is your goal here? Is it to have pre-built PDFs available for perusal when reviewing PRs? I would be mildly in favour of that. (Though I'm quite happy [fetching PRs](https://stackoverflow.com/questions/28103251/check-out-a-pull-request-on-git-to-review-test-a-pr/) and building locally.) Or is there an eventual goal to automatically commit these? I would be less keen on that, as the PDFs are of somewhat unknown third-party provenance and I think it is useful to be able to control when and which PDFs are updated.

End users look for the specs at <https://github.com/samtools/hts-specs/> about as much as they look for them at <https://samtools.github.io/hts-specs/>. So I don't think a badge is appropriate in README.md (besides for **master** it'll always be passing, right?!) and it won't interact well with _index.md_. So I would rather you added a paragraph documenting this to _MAINTAINERS.md_ and had a badge there if anywhere.

**Comment by yfarjoun** [comment]
I've been getting emails for others' branches as well....so it's probably a circle-ci thing. 

Currently, you need to log-into circle-ci in order to see the artifacts and the build...but there's an api for pulling the built artifacts if we wish to go that route. 

- I find it awkward that the pdfs are part of the repository in the first place. I think that they are like a compiled binary. They do not belong in the "source" repository, but rather they are part of a "release". I'm happy to create release scripts that find the latest artifacts from circle-ci/master and zip them up in a release in github. 

- I find it annoying that I need to download the branch in order to view the pdf (when reviewing someones PR)

- I find the process of putting a PDF in the repository annoying and error prone (Do you check that the PDF that I upload is actually the result of making the branch? Perhaps I copied the wrong pdf...) 

My end-goal is for circle-ci (or another system that someone might propose) would fix all these things:

- The PDFs do not need to be part of the repository (only part of the "release").
- The PDSs are generated for every branch automatically, so I can look at them conveniently.
- The release script copies the pdfs from the built directories and zips them into a release: we can see the script and agree on it.

**Comment by yfarjoun** [comment]
@jkbonfield the PDF's get build automatically and put in a place in circle-ci. if you log-in, you should be able to see it.

**Comment by jmarshall** [comment]
This is close to being suitable for providing pre-built PDFs to aid in reviewing PRs, if we consider that the benefit of that outweighs the onus of signing _samtools_ and ourselves up to CircleCI's service. As a basis for generating hts-specs's products (namely, legible PDF and HTML specification documents) via an automated release script, it barely scratches the surface.

This repository contains a disparate collection of unrelated specifications, each with its own cadence for updates. So there is no such thing as a “release” of the collection of PDFs — which would omit the two HTML specifications.

Nobody much likes having the PDFs in the repository, but you have to consider why they're there.

This repository (currently) contains 11 specifications written in LaTeX and two specifications written in Markdown. The **product** built from this repository is a website <<http://samtools.github.io/hts-specs/>> containing respectively PDF and HTML builds of those specifications.

That website has been advertised as the place for end users to find these specifications [since 2013](https://github.com/samtools/hts-specs/commits/2f9c96e2d26cdfdfa089eb4a4881a9f3a8d435aa). The latest edition of e.g. the SAM spec has been at <<http://samtools.github.io/hts-specs/SAMv1.pdf>> since then, and people will have bookmarked that URL.

At the time, we chose to use GitHub Pages to provide the specifications to end users because:

* it is a simple way to make a web site;
* it is neutral common ground, not privileging any organisation or institute (other than the service provider GitHub, which of course the hts-specs repo was already using) and not requiring any ongoing support or expenditure from any organisation or institute.

So the PDFs are checked in because that's how they get [copied to the web site](https://help.github.com/articles/what-is-github-pages/). Initially they were on a separate **gh-pages** branch, but the addition of Markdown-based specifications necessitated merging this into **master** (see 2a802cda6da00de1fff312051c01c562fc9983c5).

> I'm happy to create release scripts that find the latest artifacts from circle-ci/master and zip them up in a release in github.

The **purpose** of this repository is to provide these specifications to end users who wish to read them.

You seem to be suggesting that we would provide zip archives of a bunch of specifications rather than a web site of individually viewable/downloadable specifications. This would surely be less convenient for end users than the status quo. This also seems to treat the PDF and HTML specifications differently.

If the proposed GitHub releases also contained individual PDFs as artefacts, the links on the web site and in the repo's _README.md_ could point to them. However GitHub Pages provides only weak redirection facilities so the URLs would change, inconveniencing the users. And there is no way to link to the latest release's artefacts with a fixed URL, so we would need to update the links each time there was a new release, which is more work for the maintainers.

So for this proposal to be worth pursuing, you would need to explain how you propose to redesign the web site and _README.md_ to provide the specifications via a GitHub release. This would need to be no less convenient for us as maintainers (so not require extra work to update links or paste “release” notes), and there would need to be a compelling case why this is so much more convenient for us as maintainers that is it worth breaking the URLs that users have used for these specifications for the last five years.

**Comment by lbergelson** [comment]
<h3>adding demo</h3>
Demo: <strong><a href='https://circleci.com/gh/samtools/hts-specs/34/artifacts/0//home/circleci/project/demo/index.html' target='_blank'>demo</a></strong>

**Comment by lbergelson** [comment]
<h4>Here are the pdfs</h4>
<a href='https://circleci.com/gh/samtools/hts-specs/38/artifacts/0//root/project/BCFv1_qref.pdf' target='_blank'>BCFv1_qref</a>
<a href='https://circleci.com/gh/samtools/hts-specs/38/artifacts/0//root/project/BCFv2_qref.pdf' target='_blank'>BCFv2_qref</a>
<a href='https://circleci.com/gh/samtools/hts-specs/38/artifacts/0//root/project/CRAMv2.1.pdf' target='_blank'>CRAMv2.1</a>
<a href='https://circleci.com/gh/samtools/hts-specs/38/artifacts/0//root/project/CRAMv3.pdf' target='_blank'>CRAMv3</a>
<a href='https://circleci.com/gh/samtools/hts-specs/38/artifacts/0//root/project/CSIv1.pdf' target='_blank'>CSIv1</a>
<a href='https://circleci.com/gh/samtools/hts-specs/38/artifacts/0//root/project/SAMtags.pdf' target='_blank'>SAMtags</a>
<a href='https://circleci.com/gh/samtools/hts-specs/38/artifacts/0//root/project/SAMv1.pdf' target='_blank'>SAMv1</a>
<a href='https://circleci.com/gh/samtools/hts-specs/38/artifacts/0//root/project/VCFv4.1.pdf' target='_blank'>VCFv4.1</a>
<a href='https://circleci.com/gh/samtools/hts-specs/38/artifacts/0//root/project/VCFv4.2.pdf' target='_blank'>VCFv4.2</a>
<a href='https://circleci.com/gh/samtools/hts-specs/38/artifacts/0//root/project/VCFv4.3.pdf' target='_blank'>VCFv4.3</a>
<a href='https://circleci.com/gh/samtools/hts-specs/38/artifacts/0//root/project/tabix.pdf' target='_blank'>tabix</a>

**Comment by lbergelson** [comment]
<h4>Here are the pdfs</h4>
<a href='https://circleci.com/gh/samtools/hts-specs/39/artifacts/0//root/project/pdfs/BCFv1_qref.pdf' target='_blank'>BCFv1_qref</a>
<a href='https://circleci.com/gh/samtools/hts-specs/39/artifacts/0//root/project/pdfs/BCFv2_qref.pdf' target='_blank'>BCFv2_qref</a>
<a href='https://circleci.com/gh/samtools/hts-specs/39/artifacts/0//root/project/pdfs/CRAMv2.1.pdf' target='_blank'>CRAMv2.1</a>
<a href='https://circleci.com/gh/samtools/hts-specs/39/artifacts/0//root/project/pdfs/CRAMv3.pdf' target='_blank'>CRAMv3</a>
<a href='https://circleci.com/gh/samtools/hts-specs/39/artifacts/0//root/project/pdfs/CSIv1.pdf' target='_blank'>CSIv1</a>
<a href='https://circleci.com/gh/samtools/hts-specs/39/artifacts/0//root/project/pdfs/SAMtags.pdf' target='_blank'>SAMtags</a>
<a href='https://circleci.com/gh/samtools/hts-specs/39/artifacts/0//root/project/pdfs/SAMv1.pdf' target='_blank'>SAMv1</a>
<a href='https://circleci.com/gh/samtools/hts-specs/39/artifacts/0//root/project/pdfs/VCFv4.1.pdf' target='_blank'>VCFv4.1</a>
<a href='https://circleci.com/gh/samtools/hts-specs/39/artifacts/0//root/project/pdfs/VCFv4.2.pdf' target='_blank'>VCFv4.2</a>
<a href='https://circleci.com/gh/samtools/hts-specs/39/artifacts/0//root/project/pdfs/VCFv4.3.pdf' target='_blank'>VCFv4.3</a>
<a href='https://circleci.com/gh/samtools/hts-specs/39/artifacts/0//root/project/pdfs/tabix.pdf' target='_blank'>tabix</a>

**Comment by lbergelson** [comment]
<h4>Here are the pdfs</h4>

<a href='https://circleci.com/gh/samtools/hts-specs/40/artifacts/0//root/project/pdfs/SAMv1.pdf' target='_blank'>SAM v1</a>
<a href='https://circleci.com/gh/samtools/hts-specs/40/artifacts/0//root/project/pdfs/SAMtags.pdf' target='_blank'>SAM tags</a>
<p>
<a href='https://circleci.com/gh/samtools/hts-specs/40/artifacts/0//root/project/pdfs/VCFv4.3.pdf' target='_blank'>VCF v4.3</a>
<a href='https://circleci.com/gh/samtools/hts-specs/40/artifacts/0//root/project/pdfs/VCFv4.2.pdf' target='_blank'>VCF v4.2</a>
<a href='https://circleci.com/gh/samtools/hts-specs/40/artifacts/0//root/project/pdfs/VCFv4.1.pdf' target='_blank'>VCF v4.1</a>
<p>
<a href='https://circleci.com/gh/samtools/hts-specs/40/artifacts/0//root/project/pdfs/BCFv2_qref.pdf' target='_blank'>BCFv2 reference</a>
<a href='https://circleci.com/gh/samtools/hts-specs/40/artifacts/0//root/project/pdfs/BCFv1_qref.pdf' target='_blank'>BCFv1 reference</a>
<p>
<a href='https://circleci.com/gh/samtools/hts-specs/40/artifacts/0//root/project/pdfs/CRAMv3.pdf' target='_blank'>CRAM v3</a>
<a href='https://circleci.com/gh/samtools/hts-specs/40/artifacts/0//root/project/pdfs/CRAMv2.1.pdf' target='_blank'>CRAM v2.1</a>
<p>
<a href='https://circleci.com/gh/samtools/hts-specs/40/artifacts/0//root/project/pdfs/tabix.pdf' target='_blank'>Tabix</a>
<a href='https://circleci.com/gh/samtools/hts-specs/40/artifacts/0//root/project/pdfs/CSIv1.pdf' target='_blank'>CSI v1</a>

**Comment by yfarjoun** [comment]
The current implementation has a user that we control, "hts-spec-bot", provide a token (that can only comment on PRs) make a comment with the links to automatically created pdfs. this doesn't change the way pdfs are integrated into the releases. 

Does that seem like a step forward @jmarshall?

**Comment by jmarshall** [comment]
Changed PDFs as of a6918511b1c5807a51580e9de2ca5d457d08b70c: [SAMtags](
https://circleci.com/gh/samtools/hts-specs/71/artifacts/0/root/project/pdfs/SAMtags.pdf) ([diff](
https://circleci.com/gh/samtools/hts-specs/71/artifacts/0/root/project/diffs/SAMtags.pdf)), [SAMv1](
https://circleci.com/gh/samtools/hts-specs/71/artifacts/0/root/project/pdfs/SAMv1.pdf) ([diff](
https://circleci.com/gh/samtools/hts-specs/71/artifacts/0/root/project/diffs/SAMv1.pdf)), [VCFv4.2](
https://circleci.com/gh/samtools/hts-specs/71/artifacts/0/root/project/pdfs/VCFv4.2.pdf) ([diff](
https://circleci.com/gh/samtools/hts-specs/71/artifacts/0/root/project/diffs/VCFv4.2.pdf)).

**Comment by jmarshall** [comment]
When a PR is being reviewed or otherwise getting numbers of (human) comments, @hts-specs-bot's comments are noise and clutter — apart from the PDF links they contain. So I think they should be as succinct as possible, like the previous https://github.com/samtools/hts-specs/pull/328#issuecomment-422282258.

In particular for a PR like “Add new SAM OA tag”, we are all expecting that this will involve only _SAMtags.pdf_, so there is no need to spend space noting that there are no changes in CRAM/VCF/BCF/Index PDFs.

**Comment by jmarshall** [comment]
1. I wasn't expecting the diff PDFs to be worthwhile, but you have convinced me that they're useful — and useful to have as a target in the Makefile for local use. However your Makefile changes are a bit hacky and hardcoding compare-against-master is not that useful for local use, so you might consider the tidied-up version here: https://github.com/jmarshall/hts-specs/commit/diff-pdf

2. With `latexdiff` 1.2.1 I had to add `--graphics-markup none` to avoid errors making the diff PDFs for the specs that use `\includegraphcs`:

    > Cannot combine graphics markup with float styles defining \DIFaddbeginFL in terms of \DIFaddbegin. Use --graphics-markup=none option or choose a different float style

    What version of `latexdiff` is the Circle CI image using?

3. Making the CRAM diff PDFs fails with

    > ! LaTeX Error: Option clash for package ulem.

4. It would probably be more robust (and less wasteful of resources) to ask git which specs have been modified and only build those PDFs. This could be done for example via (and showing how to use the diff infrastructure in that **diff-pdf** branch):

```
mergebase=$(git merge-base HEAD master)

make --touch pdf
sleep 2  # Avoid timing issues
touch $(git diff-tree --name-only $mergebase HEAD)
# Rebuild the actually changed PDFs, and remove the merely touched ones
make pdf
find new -empty -delete

make OLD=$mergebase $(echo new/*.pdf | sed 's:new/:diff/:g')
```

5. Does CircleCI have an equivalent of Travis's `$TRAVIS_BRANCH` to get the PR's base branch? If not, **master** is usually right I guess or one could do a GitHub API call to get the PR details…

6. If a PR modifies no PDF specifications (e.g. it's a htsget PR), does the bot suppress leaving its comment?

7. Please move the scripts used only in the CircleCI stuff (_Dockerfile_ and _pdf_comment.js_) down to the _.circleci_ directory, so they're together there and so they won't be copied to the web site. What is the _Dockerfile_ for (you know, you can comment on a GitHub PR just with `curl`… :smile:), and how is all this set up? A _README[.md]_ down in this directory would be useful.

8. I still don't think there's a need for a badge on the repository home page. Unlike a fast-moving C or Java project, this build will essentially always be passing; and IMHO we should prefer not to advertise CircleCI on the hts-specs home page as it's not really relevant to hts-specs's readers.

**Comment by yfarjoun** [comment]
Thanks for all the comments @jmarshall 

CircleCi uses a docker to run all the commands in, we made a docker that will include a full tex install, javascript and curl (the js and curl are to easily make the comment without having to worry about the github api) so the Dockerfile is there to document the docker that we are using. 

I wrote up a small Readme.md file as you suggested.

Good idea about suppressing a comment if no changes to pdf. I've put that in. 

I think it would be better if the build made all the pdfs each time. I know that it "wastes" resources, but I think it's better to have a complete build at each time. I changed the comment to just point at the changes, using your elegant suggestion.

**Comment by yfarjoun** [comment]
Finally. Now I'm done changing this PR. Sorry for whoever was getting spammed by this bot.

the last spat of changes was in order to 
- enable diffs in "verbatim" environments
- make sure that there's no comment when there are no changes to pdfs.

**Comment by yfarjoun** [comment]
@jmarshall @jkbonfield can we merge this?

**Comment by jmarshall** [comment]
No, we cannot. As presented, it still contains multiple bugs.

I'm ambivalent about whether we need a _scripts_ subdirectory, but in the interests of a tidy repo and future additional scripts it's probably a good idea. So I've recently moved _genscripts.sh_ on **master** as your PR suggested. So for starters, please rebase this on current **master**, whereupon several of your _Makefile_ diffs (and the bugs therein :smile:) will evaporate…

**Comment by yfarjoun** [comment]
Thanks for the comments and fixups @jmarshall . Can you look again?

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+              make --touch diffs
+
+              # Avoid timing issues
+              sleep 2  
+
+              touch $(git diff-tree --name-only $mergebase HEAD) || true
+
+              # Rebuild the actually changed PDFs, and remove the merely touched ones
+              make diffs
+              find diffs -empty -delete -mindepth 1
```
Comment (jmarshall) [inline .circleci/config.yml]: Log file messages show that `-mindepth` is used incorrectly here.

File: CRAMv3.tex [hunk 2]
```diff
@@ -13,7 +13,6 @@
 
 \usepackage{graphicx}
 \usepackage{array}
-\usepackage{ulem}
```
Comment (jmarshall) [inline CRAMv3.tex]: The equivalent in _CRAMv2.1.tex_ has fallen off along the way. As this changes the appearance of the CRAM documents, IMHO you should make this a separate PR for the attention of the CRAM maintainers, who can decide how they want to do underlining/emphasis.

File: Makefile [hunk 3]
```diff
@@ -12,18 +12,24 @@ PDFS =	BCFv1_qref.pdf \
...
-new/SAMtags.pdf: SAMtags.tex new/SAMtags.ver
-new/VCFv4.1.pdf: VCFv4.1.tex new/VCFv4.1.ver
-new/VCFv4.2.pdf: VCFv4.2.tex new/VCFv4.2.ver
-new/VCFv4.3.pdf: VCFv4.3.tex new/VCFv4.3.ver
+new/CRAMv3.pdf:   CRAMv3.tex   new/CRAMv3.ver
+new/SAMv1.pdf:    SAMv1.tex    new/SAMv1.ver
+new/SAMtags.pdf:  SAMtags.tex  new/SAMtags.ver
+new/VCFv4.1.pdf:  VCFv4.1.tex  new/VCFv4.1.ver
+new/VCFv4.2.pdf:  VCFv4.2.tex  new/VCFv4.2.ver
+new/VCFv4.3.pdf:  VCFv4.3.tex  new/VCFv4.3.ver
```
Comment (jmarshall) [inline Makefile]: `diffs/SPEC.pdf: new/SPEC.ver` dependencies have been lost, so `make diffs` fails if you haven't already built the standards PDFs.

File: scripts/makediff.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+base_file=$1
+
+merge_base=$(git merge-base HEAD origin/master)
+
+latexdiff --version
+
+echo <<EOF > config
```
Comment (jmarshall) [inline scripts/makediff.sh]: This does nothing because `echo` should be `cat`.

If that were fixed, until a few days ago it would still do nothing due to ftilmann/latexdiff#160. I guess avoiding this bug is why your Docker image pins a particular latexdiff commit, though there is no comment or commit message explaining that. (As it happens, latexdiff 1.3.0 has been released in the last few days with this bug fixed.)

If both of those were fixed, this still does nothing because PICTUREENV takes effect within latexdiff's `sub preprocess` **after** when you would need it to.

So this _config_ stuff just plain doesn't work and you are stuck with `--graphics-markup none`.

File: .circleci/Readme.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
+We use [circle-ci](https://circleci.com/gh/samtools/hts-specs) to build the pdfs and highlight the differences for each pull-request. This is done at several steps.
+
+1. Circle-ci runs `make` and copies the pdfs and log files to an "artifacts" directory for that build. 
+2. Circle-ci runs `make diffs` on the tex files that correspond to tex files that changed since the "merge base" (defined as the most recent common ancestor of the HEAD and master branches)  
+3. Circle-ci was provided with a Token to a user (circle-ci-bot). This Token has been given permission to make comments on the github repository.
```
Comment (jmarshall) [inline .circleci/Readme.md]: The convention is _README.md_.

Presumably you mean **hts-specs-bot** here.

What I was hoping you would put in this README is details of _how all this is set up_ — if you disappeared and the CircleCI integration broke down, what would the rest of us need to do to set it up again? What samtools-related account exists on CircleCI? Who has the keys to hts-specs-bot? Exactly what permissions does it have? What are the steps to set things up? etc.

However all this may become moot as the CircleCI integration might be superseded by GitHub Actions in the coming months…

**Comment by jmarshall** [comment]
Two of the bugs added to the makefile have disappeared again during that last round of rebasing, so that is good. However there are still a number of issues with the diff stuff, and it's still not very flexible for local use.

I would suggest you use the tidying-up of the diff stuff in [this diff-pdf branch](https://github.com/jmarshall/hts-specs/commit/diff-pdf) and rebase the CircleCI stuff on top of that. (It uses a _diff/_ directory rather than _diffs/_ as that is more natural for local use: if you're working on _SAMv1.tex_ you might use `make OLD=commit diff/SAMv1.pdf` to compare to an old version, and you are comparing one specification at a time, hence singular.)

If you put _pdf_comment.js_ in the _.circleci_ directory, the CircleCI commit should just add _.circleci/*_ and not touch anything outside that. That will provide a good delineation between the CI stuff maintained by you and the PDF build/GH Pages stuff maintained by those who have been maintaining it.

And especially for newly-added files, there are the usual comments about trailing whitespace on lines (please don't) and a newline at the end of the file (please do, otherwise git will say _No newline at end of file_ for ever more).

**Comment by lbergelson** [comment]
@yfarjoun For you to add where you think it should go:


## All about the hts-specs-bot:

The [hts-specs-bot](https://github.com/hts-specs-bot) is a github account used to comment on hts-specs pull requests.

## If a key needs to be rotated:
-  Sign into the bot account. You can do this in an incognito window to avoid signing out of your existing github account.
   -    username: `hts-specs-bot`
   -  password:  *this is a secret and I'm not sure where to put it... possibly in a google doc, a private samtools repo or as a comment in the hts-specs-bot google group*
-  Navigate to Settings > Developer Settings > Personal access tokens
-  Delete the existing token
-  Generate a new token with the `public_repo` scope.
    - In addition to allowing the bot to make comments on pull requests, this scope allows write access to all public repos that the bot account can access.  For that reason it is **very important** not to give this bot write access to anything.  
-  Update the existing token in circle-ci by navigating to [Settings -> Environment Variables](https://circleci.com/gh/samtools/hts-specs/edit#env-vars) and replacing the existing token there with the newly generated one.

### Bot Password
If the bot account password is lost it can be recovered by sending a recovery email to hts-specs-bot@broadinstitute.org.  Spec maintainers should be made owners of that [private google group](https://groups.google.com/a/broadinstitute.org/forum/#!forum/hts-specs-bot)

### Circle-CI docker image
The docker image used by the Circle-CI build is hosted on docker under [htsspecs/circle-ci-image](https://hub.docker.com/r/htsspecs/circle-ci-image/).  Spec maintianers should be made owners of that repository.

File: CRAMv3.tex (hunk 2 above)
Comment (yfarjoun) [reply CRAMv3.tex]: I changed this since it didn't work well with the `latexdiff`....and it didn't seem to change the appearance that much....

**Comment by yfarjoun** [comment]
@jmarshall, your make-fu is truly strong 🙇 

I copied most of you suggestions from your branch, but I didn't make the new/pdfs for each diff/pdf. I was unclear of the advantage there. 

Please let me know if the current state is satisfactory, and sorry for all the emails.... :-)

File: .circleci/config.yml (hunk 1 above)
Comment (yfarjoun) [reply .circleci/config.yml]: fixed

File: CRAMv3.tex (hunk 2 above)
Comment (yfarjoun) [reply CRAMv3.tex]: I think that we concluded that this is OK in the call.

**Comment by yfarjoun** [comment]
@jmarshall, I think that this is waiting for your final review.

File: CRAMv3.tex (hunk 2 above)
Comment (jkbonfield) [reply CRAMv3.tex]: We accepted in the call that there is room for changing which packages are used as it was a rather arbitrary decision made by some ghastly xslt docx to tex conversion script.  It wasn't clear what the precise effect would be though, so we'd want to evaluate.

If it's just for making \emph underlined (which I must admit I was somewhat surprised to discover it did that in CRAM doc but hadn't got around to looking for out what was causing it), then I don't care about removing it.

Comment (jkbonfield) [reply CRAMv3.tex]: Update: I find my old copy of CRAMv2.1.docx and sure enough it has underline for those bits of texts.  I expect the conversion script was just trying to faithfully reproduce the same formatting, but I'm happy to be more traditional and use italics in such cases.

Comment (jmarshall) [reply CRAMv3.tex]: I've gone through all the \emphs in the two CRAM documents and have a PR coming soon to do something appropriate for each of them and then nuke ulem.

Comment (jkbonfield) [reply CRAMv3.tex]: Thanks.  I see one of them at least is a citation.  Do we have a recommended way of doing this?  I spotted the Peter Cock one in SAMv1.tex, but that is also just simulated via combinations of textit and emph.  I've done something similar for my fledgeling CRAMv4.tex document.

Comment (jmarshall) [reply CRAMv3.tex]: There's several in the SAM specs, and they use `{\bf }{\sl }\href{URL}{doi:…}` etc to hand-roll a standard citation format (as a footnote).

The current draft of the crypt4gh spec uses bibtex with full information on the referenced works in a _references.bib_ file. If that ever lands in hts-specs I might be tempted to convert to that method and join them.

Comment (jkbonfield) [reply CRAMv3.tex]: I wouldn't object to going down the bibtex route, but happy to go with the current flow until it becomes a major concern.

Comment (jmarshall) [reply CRAMv3.tex]: This is now PR #373.

**Comment by yfarjoun** [comment]
I rebased and squashed into one commit. hopefully we can move ahead with this...

**Comment by jmarshall** [comment]
1. As noted twice before, please move _pdf_comment.js_ to the _.circleci_ directory. That will provide a clear delineation between the CI stuff maintained by you and the PDF build/GH Pages stuff maintained by those who have been maintaining it.

1. Your Dockerfile builds its own latexdiff as of a random commit between releases 1.2.1 and the then-upcoming 1.3.0. Release 1.3.0 has been available since October and fixes the bug you're pinning latexdiff to avoid; but in fact you are no longer using `latexdiff -c` so that bug fix is immaterial. So you might as well return to using the base blang/latex:ubuntu latexdiff or at least build the new release rather than some random commit.

1. The list of files in _.circleci/README.md_ needs updating.

1. I don't understand what you mean by “I didn't make the new/pdfs for each diff/pdf”. Your `diffs/%.pdf: %.tex new/%.pdf` rule in fact **does** make the new/pdf for each diff/pdf -- unnecessarily, as all that is needed is the new/ver. And your `new/%.pdf: %.tex new/%.ver` rule makes every PDF build a .ver file first, even the ones that don't use a .ver file. Especially the latter is not the end of the world, but as it is simple to express the dependencies exactly it would be preferable to do so — which is what the existing rules and those in the [diff-pdf](
https://github.com/jmarshall/hts-specs/commit/diff-pdf) branch do.

1. As noted before, I would prefer the directory and targets to be “diff” rather than “diffs”. The other makefile targets (all, pdf) and directories (new/) are all singular. And that is more natural for local use: if you're working on _SAMv1.tex_ you might use `make OLD=commit diff/SAMv1.pdf` to compare to an old version, and you are comparing one specification at a time, hence singular.

1. If you're adding a phony target, add it to `.PHONY`.

1. In the clean rule, `-rm -f` avoids spurious error messages when the files already don't exist.

1. Please avoid incidental whitespace changes in otherwise unchanged files or parts of files, as in _Makefile_, _README.md_, __config.yml_, and, non-whitespace-wise, _CRAMv3.tex_. Those double blank lines between different sections of the makefile are entirely intentional.

1. The audience for the top-level _README.md_ (aka index.html on the web site) is people referring to the specifications. Circle CI is of no interest to them, so we don't really need anything in the README about it.

1. As noted before, especially for newly-added files, there are the usual comments about trailing whitespace on lines (please don't) and a newline at the end of the file (please do, otherwise git will say _No newline at end of file_ for ever more).

1. The commit message needs updating.

**Comment by jmarshall** [comment]
Also please remove the “Rob Ot” name from the hts-specs-bot user. To be sure it's cute, but email notifications from Rob Ot are disconcerting (Rob who?!), while notifications from hts-specs-bot would be clearly mechanical.

**Comment by jmarshall** [comment]
Perhaps contrary to appearances :smile: I am very keen to get this merged.

One way to expedite this would be, once some agreement has been reached on (5) _diff_ vs _diffs_, to merge that [diff-pdf branch](
https://github.com/jmarshall/hts-specs/commit/diff-pdf) to master and rebase the CI additions on top of that — or equivalently to rebase the CI additions on top of that commit so that that diff-pdf commit becomes part of this PR.

That should at a stroke resolve concerns 4, 6, 7, 8, and 9. I've updated that **diff-pdf** branch to provide a phony target as used by _.circleci/config.yml_ and follow your version's `latexdiff-vc` command exactly.

File: .circleci/config.yml (hunk 1 above)
Comment (jmarshall) [reply .circleci/config.yml]: Not fixed. It still always logs a warning:

> find: warning: you have specified the -mindepth option after a non-option argument -empty, but options are not positional (-mindepth affects tests specified before it as well as those specified after it).  Please specify options before other arguments.

**Comment by yfarjoun** [comment]
@jmarshall I think I addressed all your issues....but I probably missed a few :-)

**Comment by yfarjoun** [comment]
@jmarshall I know it's been a long slog...I hope I can encourage you to bless this version...

**Comment by jmarshall** [comment]
Rebased and tidied up the trailing whitespace. Also I realised that for the makefile targets we can just have both `diff` and `diffs` (like htslib & samtools have both `test` and `check` targets) so everybody can be happy in their _make_ commands…

**Comment by jmarshall** [comment]
And finally merged.   🎉   Thanks for all your work on this, Yossi!

[I deleted the **yf_add_circle_ci** branch immediately as it pointed to the same commit as **master**, which could have an impact on fresh clones.]
